Introduce idea of a TimingFunction

jacksontaylor13 opened the pull request:
Given by issue #77, we introduce this idea of a TimingFunction now. This should allow any sort function to be considered a continuous or inter item function. This separates the notion that the two are related and brings TimingFunction to be an entirely new concept.

- [ ] Update changelog

jacksontaylor13 committed:
Introduce idea of a TimingFunction

Given by issue #77, we introduce this idea of a TimingFunction now. This should allow any sort function to be considered a continuous or inter item function. This separates the notion that the two are related and brings TimingFunction to be an entirely new concept.

jacksontaylor13 committed:
Update code to support new timing functions

With the new `TimingFunction` protocol being added, we needed to make sure that it works properly with the example app. This will be a breaking change for some functions but we are going to submit another PR on top of this that should minimize the breaking nature of this change.

jacksontaylor13 commented:
Still need to fix the unit tests, since the tests currently rely on `SortFunction`'s returning a `TimedView`

ianterrell commented on Sources/Classes/Timing Functions/DurationTimingFunction.swift:
```diff
@@ -0,0 +1,36 @@
+        
+    }
+    
+    public init(duration: TimeInterval) {
```
Rather than the default on the stored property you could consider a default value for this parameter and remove the empty `init` above.

ianterrell commented on Sources/Classes/Timing Functions/InterItemTimingFunction.swift:
```diff
@@ -0,0 +1,40 @@
+        
+    }
+    
+    public init(interObjectDelay: TimeInterval) {
```
Ditto here with default parameter.

ianterrell commented on Sources/Classes/Full View/FullViewAnimation.swift:
```diff
@@ -52,8 +52,9 @@ public extension Spruce {
     ///   - completion: a closure that is called upon the final animation completing. A `Bool` is passed into the closure letting you know if the animation has completed. **Note:** If you stop animations on the whole animating view, then `false` will be passed into the completion closure. However, if the final animation is allowed to proceed then `true` will be the value passed into the completio ...
-    public func animate(withSortFunction sortFunction: SortFunction, prepare: PrepareHandler? = nil, animation: Animation, exclude: [UIView]? = nil, recursiveDepth: Int = 0, completion: CompletionHandler? = nil) {
-        var timedViews = sortFunction.timeOffsets(view: self.view, recursiveDepth: recursiveDepth)
+    public func animate(withSortFunction sortFunction: SortFunction, timingFunction: TimingFunction, prepare: PrepareHandler? = nil, animation: Animation, exclude: [UIView]? = nil, recursiveDepth: Int = 0, completion: CompletionHandler? = nil) {
```
The curse of good documentation — you need to update the parameter above.

ianterrell commented on Sources/Classes/Sort Functions/ContinuousWeightedSortFunction.swift:
```diff
@@ -45,41 +45,26 @@ public struct ContinuousWeightedSortFunction: PositionSortFunction, WeightSortFu
-
-    public func timeOffsets(view: UIView, recursiveDepth: Int) -> [TimedView] {
+    
+    public func weights(forView view: UIView, recursiveDepth: Int) -> [WeightedView] {
```
`weights(for view: UIView...` is probably slightly more idiomatic naming.

ianterrell commented on Sources/Classes/Timing Functions/TimingFunction.swift:
```diff
@@ -0,0 +1,30 @@
+
+import Foundation
+
+/// An internal struct that is used to relate a view with a time offset. This is used to determine when each view is set to animate.
```
The tiniest bit confusing to call it an "internal struct".

ianterrell reviewed (commented):
I only really gave it a cursory skim, but: I'm glad to see this change making it in! I think it's a good improvement in the concepts of an already great library. 👍

jacksontaylor13 committed:
Add stock timing functions (#84)

Add stock timing functions

To make calling a timing function easier, there will now be an `enum` called `StockTimingFunction` that uses associated values to make creating a timing function super quick and easy.

jacksontaylor13 committed:
Merge branch 'master' into feature/sort-function-weighting

ci-wta commented:
<!--
  1 Error: Any changes to the Spruce libr...
  1 Warning: Big PR
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Any changes to the Spruce library need to be documented in the Changelog</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>
